core/templates/dev/head/expressions/expressionSyntaxTree.js

```diff
@@ -0,0 +1,420 @@
+        throw new ExprWrongArgTypeError(
+          originalValue, typeof originalValue, 'Number');
+      };
+
+      // Coerces all values in the given argument array to Number, and throws
+      // an error if the result is NaN.
+      var _coerceAllArgsToNumber = function(args) {
+        for (var i = 0; i < args.length; i++) {
+          args[i] = _coerceToNumber(args[i]);
+        }
+        return args;
+      };
+
+      // NOTE TO DEVELOPERS: When adding a new reserved word to this object,
+      //   please first ensure that existing explorations do not use this
+      //   parameter name. Also, to prevent future explorations using it,
+      //   modify feconf.INVALID_PARAMETER_NAMES accordingly.
+      // TODO(kashida): Document all operators input and output contracts.
+
+      var system = {
+    '#': {
+      type: function(args, envs) {
+        return lookupEnvs(args + '', envs);
+      },
+      eval: function(args, envs) {
+        return lookupEnvs(args[0] + '', envs);
+      }
+    },
+    '+': {
+      type: function(args, envs) {
+        verifyNumArgs(args, 1, 2);
+        _verifyArgsHaveType(args, 'Real');
+        return 'Real';
+      },
+      eval: function(args, envs) {
+        verifyNumArgs(args, 1, 2);
+        var numericArgs = _coerceAllArgsToNumber(args);
+        return numericArgs.length == 1 ? numericArgs[0] :
+          numericArgs[0] + numericArgs[1];
+      }
+    },
+    '-': {
+      type: function(args, envs) {
+        verifyNumArgs(args, 1, 2);
+        _verifyArgsHaveType(args, 'Real');
+        return 'Real';
+      },
+      eval: function(args, envs) {
+        verifyNumArgs(args, 1, 2);
+        var numericArgs = _coerceAllArgsToNumber(args);
+        return numericArgs.length == 1 ? -numericArgs[0] :
+          numericArgs[0] - numericArgs[1];
+      }
+    },
+    '*': {
+      type: function(args, envs) {
+        verifyNumArgs(args, 2);
+        _verifyArgsHaveType(args, 'Real');
+        return 'Real';
+      },
+      eval: function(args, envs) {
+        verifyNumArgs(args, 2);
+        var numericArgs = _coerceAllArgsToNumber(args);
+        return numericArgs[0] * numericArgs[1];
+      }
+    },
+    '/': {
+      type: function(args, envs) {
+           verifyNumArgs(args, 2);
+           _verifyArgsHaveType(args, 'Real');
+           return 'Real';
+         },
+      eval: function(args, envs) {
+           verifyNumArgs(args, 2);
+           var numericArgs = _coerceAllArgsToNumber(args);
+           return numericArgs[0] / numericArgs[1];
+         }
+    },
+    '%': {
+      type: function(args, envs) {
+           verifyNumArgs(args, 2);
+           _verifyArgsHaveType(args, 'Real');
+           return 'Real';
+         },
+      eval: function(args, envs) {
+           verifyNumArgs(args, 2);
+           var numericArgs = _coerceAllArgsToNumber(args);
+           return numericArgs[0] % numericArgs[1];
+         }
+    },
+    '<=': {
+      type: function(args, envs) {
```

Comment:
The reason I limited it to String and Real is because the main goal here is to allow for the UI to allow only the appropriate type of parameters into the rules. So more or less anything that isn't a number is considered a string at this stage. I'm not sure that there is a reason at this point to have more types than just that, because marking it as a "Boolean" doesn't give it any new kind of distinction within the code. If there is ever a point where it being boolean means something, it could make sense to make that distinction. Right now isn't that point.